[WIP] Forms: Create new base blocks to enable Global Styles support JETPACK-261

aaronrobertshaw opened the pull request:
JETPACK-261

Fixes #42355

⚠️ **Please note: This PR forms the basis for migrating form fields to inner blocks. It should not be merged until the [related PR](https://github.com/Automattic/jetpack/pull/42890) migrating to these new blocks proves they are sufficiently complete**. ⚠️ 

## Proposed changes:

* Add news blocks for greater control over styling forms and to support global styles
- New blocks include:
    - `jetpack/input`
    - `jetpack/label`
    - `jetpack/option`
    - `jetpack/options`

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:

TBA

aaronrobertshaw committed:
Add base form field blocks

Introduces foundational form field blocks:
- jetpack/input
- jetpack/label
- jetpack/option
- jetpack/options

These blocks will be reused across all field types using inner blocks.

aaronrobertshaw committed:
changelog

aaronrobertshaw committed:
Fix issues with style syncing hook:

- Prevents adding bulk changes when syncing blocks to the undo stack
- Correctly determines the parent field block for option blocks to check
  if syncing is enabled
- Correctly find all option blocks regardless of whether they are a
  field's direct child, e.g. consent, or nested within an options block
  e.g. single/multiple choice fields.

aaronrobertshaw committed:
Prevent style syncing when blocks are reused in previews

andrewserong commented on projects/packages/forms/src/blocks/option/index.js:
```diff
@@ -0,0 +1,83 @@
+			type: 'string',
+			default: '',
+		},
+		// TODO: Work out the best approach here for type.
+		// Currently, the type attribute isn't required as it can be passed via block
+		// context. It turns out though that choice items for the single and multiple
+		// choice fields had custom and different icons. Sharing a single
+		// `jetpack/option` block for global styling purposes means that can't happen
+		// without block variations. Block variations would require the type attribute
+		// at this block's level to allow different icons for each variation.
+		//
+		// Is this really needed?
+		//
+		// type: {
+		// 	type: 'string',
+		// 	default: 'checkbox',
+		// },
```
Is this just about different icons within the editor? If that's the only difference between the two, then I'd lean toward not including a `type` attribute for now, and circle back to it if we wind up feeling like there's a need for it further down the track.

aaronrobertshaw replied:
It is. 

This came up in discussion with @simison and the expectation is we might ideally have separate icons. The catch is we'll need to refactor this attribute sooner rather than later to avoid needing deprecations if we waited until post-release of inner blocks.

aaronrobertshaw committed:
Ensure label is editable after being emptied and placeholder shows

aaronrobertshaw committed:
Opt out of block gap spacing for Options block

This is an enhancement over trunk however the is some work required to
make this actually work, especially given the shortcode rendering on the
frontend. For the time being this support will explicitly be opted out
of. It can be revisited when the benefit justifies the additional
complexity and work.

aaronrobertshaw commented:
With many hands working on testing and refining #42890, it's proven tricky to ensure this base branch is kept up-to-date with all the tweaks to the new base blocks without interfering with progress and momentum on #42890.

After some discussion with @talldan, I've decided to close this PR so we only have a single PR to manage with #42890 now to be based off trunk.
